pkg/validation/validation_test.go

```diff
@@ -0,0 +1,480 @@
+					"one": {
+						Dependencies: []string{"two"},
+					},
+					"two": {
+						Dependencies: []string{"one"},
+					},
+				},
+				Selector: &k8sApiUnv.LabelSelector{
+					MatchLabels: map[string]string{"a": "b"},
+				},
+			},
+		},
+		"spec.steps: Forbidden: detected cycle [three four five]": {
+			ObjectMeta: k8sApi.ObjectMeta{
+				Name:      "mydag",
+				Namespace: k8sApi.NamespaceDefault,
+				UID:       k8sTypes.UID("1uid1cafe"),
+			},
+			Spec: api.WorkflowSpec{
+				Steps: map[string]api.WorkflowStep{
+					"one": {},
+					"two": {
+						Dependencies: []string{"one"},
+					},
+					"three": {
+						Dependencies: []string{"two", "five"},
+					},
+					"four": {
+						Dependencies: []string{"three"},
+					},
+					"five": {
+						Dependencies: []string{"four"},
+					},
+					"six": {
+						Dependencies: []string{"five"},
+					},
+				},
+				Selector: &k8sApiUnv.LabelSelector{
+					MatchLabels: map[string]string{"a": "b"},
+				},
+			},
+		},
+		"spec.steps: Not found: \"three\"": {
+			ObjectMeta: k8sApi.ObjectMeta{
+				Name:      "mydag",
+				Namespace: k8sApi.NamespaceDefault,
+				UID:       k8sTypes.UID("1uid1cafe"),
+			},
+			Spec: api.WorkflowSpec{
+				Steps: map[string]api.WorkflowStep{
+					"one": {},
+					"two": {
+						Dependencies: []string{"three"},
+					},
+				},
+				Selector: &k8sApiUnv.LabelSelector{
+					MatchLabels: map[string]string{"a": "b"},
+				},
+			},
+		},
+		"spec.activeDeadlineSeconds: Invalid value: -42: must be greater than or equal to 0": {
+			ObjectMeta: k8sApi.ObjectMeta{
+				Name:      "mydag",
+				Namespace: k8sApi.NamespaceDefault,
+				UID:       k8sTypes.UID("1uid1cafe"),
+			},
+			Spec: api.WorkflowSpec{
+				Steps: map[string]api.WorkflowStep{
+					"one": {},
+				},
+				ActiveDeadlineSeconds: &negative64,
+				Selector: &k8sApiUnv.LabelSelector{
+					MatchLabels: map[string]string{"a": "b"},
+				},
+			},
+		},
+		"spec.selector: Required value": {
+			ObjectMeta: k8sApi.ObjectMeta{
+				Name:      "mydag",
+				Namespace: k8sApi.NamespaceDefault,
+				UID:       k8sTypes.UID("1uid1cafe"),
+			},
+			Spec: api.WorkflowSpec{
+				Steps: map[string]api.WorkflowStep{
+					"one": {},
+				},
+			},
+		},
+		"spec.steps: Invalid value: \"one\": jobTemplate and externalRef are mutually exclusive": {
+			ObjectMeta: k8sApi.ObjectMeta{
+				Name:      "mydag",
+				Namespace: k8sApi.NamespaceDefault,
+				UID:       k8sTypes.UID("1uid1cafe"),
+			},
+			Spec: api.WorkflowSpec{
+				Steps: map[string]api.WorkflowStep{
+					"one": {
+						JobTemplate: &k8sBatch.JobTemplateSpec{},
+						ExternalRef: &k8sApi.ObjectReference{},
+					},
+				},
+				Selector: &k8sApiUnv.LabelSelector{
+					MatchLabels: map[string]string{"a": "b"},
+				},
+			},
+		},
+	}
+
+	for k, v := range errorCases {
+		errs := ValidateWorkflow(&v)
+		if len(errs) == 0 {
+			t.Errorf("expected failure for %s", k)
+		} else {
+			s := strings.Split(k, ":")
```

Comment:
This seems to cause undesired behavior. For example: the errorCase specified on line 308 contains multiple double-colons, such that only "Invalid Value" will be matched against the error message, instead of the entire expected error message.